cpp/src/io/orc/reader_impl.cu

```diff
@@ -472,92 +734,113 @@ table_with_metadata reader::impl::read(size_type skip_rows,
+              "Unexpected discrepancy in bytes read.");
+          } else {
+            const auto buffer =
+              _metadata->per_file_metadata[stripe_source_mapping.source_idx].source->host_read(
+                offset, len);
+            CUDF_EXPECTS(buffer->size() == len, "Unexpected discrepancy in bytes read.");
+            CUDA_TRY(
+              cudaMemcpyAsync(d_dst, buffer->data(), len, cudaMemcpyHostToDevice, stream.value()));
+            stream.synchronize();
+          }
         }
-        if (_source->is_device_read_preferred(len)) {
-          CUDF_EXPECTS(_source->device_read(offset, len, d_dst, stream) == len,
-                       "Unexpected discrepancy in bytes read.");
-        } else {
-          const auto buffer = _source->host_read(offset, len);
-          CUDF_EXPECTS(buffer->size() == len, "Unexpected discrepancy in bytes read.");
-          CUDA_TRY(
-            cudaMemcpyAsync(d_dst, buffer->data(), len, cudaMemcpyHostToDevice, stream.value()));
-          stream.synchronize();
-        }
-      }
 
-      // Update chunks to reference streams pointers
-      for (size_t j = 0; j < num_columns; j++) {
-        auto &chunk         = chunks[i * num_columns + j];
-        chunk.start_row     = stripe_start_row;
-        chunk.num_rows      = stripe_info->numberOfRows;
-        chunk.encoding_kind = stripe_footer->columns[_selected_columns[j]].kind;
-        chunk.type_kind     = _metadata->ff.types[_selected_columns[j]].kind;
-        chunk.decimal_scale = _metadata->ff.types[_selected_columns[j]].scale.value_or(0);
-        chunk.rowgroup_id   = num_rowgroups;
-        chunk.dtype_len     = (column_types[j].id() == type_id::STRING)
-                            ? sizeof(std::pair<const char *, size_t>)
-                            : cudf::size_of(column_types[j]);
-        if (chunk.type_kind == orc::TIMESTAMP) {
-          chunk.ts_clock_rate = to_clockrate(_timestamp_type.id());
+        // Update chunks to reference streams pointers
+        for (size_t col_idx = 0; col_idx < num_columns; col_idx++) {
+          auto &chunk         = chunks[stripe_chunk_index * num_columns + col_idx];
+          chunk.start_row     = stripe_start_row;
+          chunk.num_rows      = stripe_info->numberOfRows;
+          chunk.encoding_kind = stripe_footer->columns[_selected_columns[col_idx]].kind;
+          chunk.type_kind     = _metadata->per_file_metadata[stripe_source_mapping.source_idx]
+                              .ff.types[_selected_columns[col_idx]]
+                              .kind;
+          chunk.decimal_scale = _metadata->per_file_metadata[stripe_source_mapping.source_idx]
+                                  .ff.types[_selected_columns[col_idx]]
+                                  .scale.value_or(0);
+          chunk.rowgroup_id = num_rowgroups;
+          chunk.dtype_len   = (column_types[col_idx].id() == type_id::STRING)
+                              ? sizeof(std::pair<const char *, size_t>)
```

Comment:
should this be 
```suggestion
                              ? sizeof(thrust::pair<const char *, size_type>)
```
@rgsl888prabhu ?

Reply:
Yes, that should be `sizeof(thrust::pair<const char *, size_type>)` or may be we can use `string_index_pair`